digitaltwin_client/src/dt_client_core.c

```diff
@@ -849,21 +779,23 @@ DIGITALTWIN_CLIENT_RESULT DT_ClientCoreRegisterInterfacesAsync(DT_CLIENT_CORE_HA
     {
         LogError("DT_InterfaceList_BindInterfaces failed, result = %d", result);
     }
-    else if ((result = SendInterfacesToRegisterMessage(dtClientCore, deviceCapabilityModel)) != DIGITALTWIN_CLIENT_OK)
-    {
-        LogError("SendInterfacesToRegisterMessage failed");
-    }
     else
     {
         dtClientCore->registerInterfacesCallbackContext.dtInterfaceRegisteredCallback = dtInterfaceRegisteredCallback;
         dtClientCore->registerInterfacesCallbackContext.userContextCallback = userContextCallback;
-    
+
         // Once we're processing an interface update, no other caller initiated operations on this device client can occur.
         dtClientCore->registrationStatus = DT_CLIENT_REGISTRATION_STATUS_REGISTERING;
         result = DIGITALTWIN_CLIENT_OK;
     }
 
     (void)InvokeBindingUnlock(dtClientCore, &lockHeld);
+
+    if (result == DIGITALTWIN_CLIENT_OK)
+    {
+        InvokeBindingOnDemand(dtClientCore);
```

Comment:
>InvokeBindingOnDemand [](start = 8, length = 21)

#1
This needs to be called while the BindingLock is still held.  So it'd be something like

else if (result=DT_Interface_BindInterfaces()...) 
{
  // error
}
else if (result = InvokeBindingOnDemand() ...)
{

}

#2
Instead of InvokeBindingOnDemand(), maybe something like SubscribeForIncomingRequests() or something.

#3
SetRegistrationCompleteInvokeCallbacks() however needs to not be called with the lock.  Here I recommenda  new helper function, *only* called on success of everything here of course, and that does do the callback stuff.

Something like

else if (...) // blah, core things
{
...
}
(void)InvokeBindingUnlock(dtClientCore, &lockHeld);

    if (result == DIGITALTWIN_CLIENT_OK)
    {
        SetRegistrationCompleteInvokeCallbacks(dtClientCore);
    }

And in then new SetRegistrationCompleteInvokeCallbacks(), you'll need to do the BeginClientCoreCallbackProcessing() and EndClientCoreCallbackProcessing().  These make sure that handle can't be killed under us while we invoke a callback routiune, but at same time the lock for dt_client_core is NOT held so we do'nt run into deadlocks.